test/functional/feature_checktemplateverify.py

```diff
@@ -0,0 +1,289 @@
+    """
+    random_address = script_to_p2sh(CScript())
+    rawtx = node.createrawtransaction(inputs=[{"txid": txid, "vout": 0}], outputs={random_address: amount})
+    tx = CTransaction()
+    tx.deserialize(BytesIO(hex_str_to_bytes(rawtx)))
+    # Replace with our script
+    tx.vout[0].scriptPubKey = script
+    # Sign
+    signresult = node.signrawtransactionwithwallet(tx.serialize().hex())
+    assert_equal(signresult["complete"], True)
+    raw_tx =  signresult['hex']
+    tx = CTransaction()
+    tx.deserialize(BytesIO(hex_str_to_bytes(raw_tx)))
+    return tx
+
+class CheckTemplateVerifyTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [['-whitelist=127.0.0.1', '-par=1']]  # Use only one script thread to get the exact reject reason for testing
+        self.setup_clean_chain = True
+        self.rpc_timeout = 120
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+    def get_block(self, txs):
+        self.tip    = self.nodes[0].getbestblockhash()
+        self.height = self.nodes[0].getblockcount()
+        block = create_block(int(self.tip, 16), create_coinbase(self.height))
+        block.vtx.extend(txs)
+        add_witness_commitment(block)
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        return block.serialize(True).hex(), block.hash
+    def add_block(self, txs):
+        block, h = self.get_block(txs)
+        self.nodes[0].submitblock(block)
+        assert_equal(self.nodes[0].getbestblockhash(), h)
+        return h
+    def fail_block(self, txs, cause = CHECKTEMPLATEVERIFY_ERROR):
+        block, h = self.get_block(txs)
+        assert_equal(self.nodes[0].submitblock(block), cause)
+        assert_equal(self.nodes[0].getbestblockhash(), self.tip)
+
+    def run_test(self):
+
+        # The goal is to test a number of circumstances and combinations of parameters. Roughly:
+        #
+        #   - SegWit OP_CTV
+        #   - Bare OP_CTV
+        #   - OP_CTV at vin index 0
+        #   - OP_CTV at vin index > 0
+        #   - OP_CTV with scriptSigs set
+        #   - OP_CTV without scriptSigs set
+        #   - OP_CTV with multiple inputs
+        #   - accepting correct parameters
+        #   - rejecting incorrect parameters
+        #   - OP_CTV in a tree
+        #
+        # A few tests may seem redundant, but it is because they are testing the cached computation of the hash
+        # at vin index 0
+
+        self.nodes[0].add_p2p_connection(P2PInterface())
+
+        BLOCKS = 110
+        self.log.info("Mining %d blocks for mature coinbases", BLOCKS)
+        self.coinbase_txids = [self.nodes[0].getblock(b)['tx'][0] for b in self.nodes[0].generate(BLOCKS)]
+
+        self.log.info("Creating setup transactions")
+        outputs, script = random_real_outputs_and_script(10)
+        # Add some fee satoshis
+        amount = (sum(out.nValue for out in outputs)+200*500) /100e6
```

Comment:
I know this is just the tests, but it would be nice to not use floating point

Reply:
100e6 refers to the COIN constant?

Reply:
yeah I think it kinda has to be floating point because the RPCs use BTC and we have sats... if I used a non float I would truncate?

But I can refactor it to be COIN? Or maybe preserve the values as sats for longer and only convert to BTC in one place?

Reply:
There is decimal.Decimal